### PR TITLE
chore(central): simplify v2 and v4 handling

### PR DESCRIPTION
**Pull request by jvdm**
## Description

Simplifies the scanner definitions handler by combining related functions for Scanner v2 and V4. This also removes an annoying misleading log about missing offline files, when they're not needed.

I also removed `startUpdaterAndOpenFile` because it's no longer used.

This is the first PR in a series of PRs to fix this file for bugs and readability purposes:

* https://github.com/stackrox/stackrox/pull/11212 <-- This
* https://github.com/stackrox/stackrox/pull/11825 (replaces https://github.com/stackrox/stackrox/pull/11282 which was originally merged into here but was somehow lost...)
* https://github.com/stackrox/stackrox/pull/11836
* https://github.com/stackrox/stackrox/pull/11837
* https://github.com/stackrox/stackrox/pull/11841
* https://github.com/stackrox/stackrox/pull/11842
* https://github.com/stackrox/stackrox/pull/11843

Once the followups are merged into this branch, then this branch will be merged.

## Checklist
- [x] Investigated and inspected CI test results
- ~[ ] Unit test and regression tests added~
- ~[ ] Evaluated and added CHANGELOG entry if required~
- ~[ ] Determined and documented upgrade steps~
- ~[ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))~

### Here I tell how I validated my change

This was tested in https://github.com/stackrox/stackrox/pull/11283.

### Reminder for reviewers

In addition to reviewing code here, reviewers **must** also review testing and request further testing in case the
performed one does not seem sufficient. As a reviewer, you must not approve the change until you understand the
performed testing and you are satisfied with it.

### CODE REVIEW

**Comment by RTann** [comment]
I created https://github.com/stackrox/stackrox/pull/11246 built on top of this to resolve other concerns I had

File: central/scannerdefinitions/handler/handler.go
Comment (RTann) [inline central/scannerdefinitions/handler/handler.go:538]: Thanks for fixing this. The log was misleading

**Review by RTann — APPROVED** [review]
I'm ok with all of this, as I have built more on top of this PR. Not sure if we'd want to merge all of them into here first then merge or merge this then the rest.

See https://github.com/stackrox/stackrox/pull/11282 and https://github.com/stackrox/stackrox/pull/11283

**Review by RTann — CHANGES_REQUESTED** [review]
Blocking until both followup PRs are merged into this one.

File: central/scannerdefinitions/handler/handler.go
Comment (dcaravel) [inline central/scannerdefinitions/handler/handler.go:451]: This looks like will always vail for the V2 use case, `fileName` will always be the empty string, has this been tested?

File: central/scannerdefinitions/handler/handler.go
Comment (dcaravel) [inline central/scannerdefinitions/handler/handler.go:421]: Doesn't look like `v2UpdaterType` belongs here.  Tested this out with the image from and V2 is broken:


```
{"Event":"Updater failed","Level":"error","Location":"updater.go:86","Time":"2024-07-23 20:11:31.329024","error":"fetching update from URL: received response code 500 Internal Server Error, but expected 2xx; error message: {\"code\":13,\"message\":\"could not read vulnerability definition fb5a20c3-7c89-42da-a7ef-4678261a096f: extracting: open : invalid argument\"}"}
```

**Review by dcaravel — CHANGES_REQUESTED** [review]
This breaks the V2 use case / will always fail.  Test cases do not appear to cover the scenarios which break.

Would recommend testing this for both Scanner V2 and V4 on a fresh install in online and offline mode and confirm definitions are pulled successfully.

File: central/scannerdefinitions/handler/handler.go
Comment (RTann) [reply central/scannerdefinitions/handler/handler.go:372]: I believe I address this in the followup PRs), which makes it harder to verify unfortunately

File: central/scannerdefinitions/handler/handler.go
Comment (RTann) [reply central/scannerdefinitions/handler/handler.go:463]: (hopefully) fixed this in a recent commit

File: central/scannerdefinitions/handler/handler.go
Comment (dcaravel) [inline central/scannerdefinitions/handler/handler.go:428]: nit: add a note next to this defer to indicate why `openedFile` is closed, something like "`openFromArchive` creates a new file handle, because it does not use the `openedFile` handle we close it to avoid a leak"

File: central/scannerdefinitions/handler/handler_test.go
Comment (dcaravel) [inline central/scannerdefinitions/handler/handler_test.go:403]:
```suggestion
	w = httptest.NewRecorder()
	req.Header.Set("X-Scanner-V4-Accept", "application/vnd.stackrox.scanner-v4.multi-bundle+zip")
```
Adding this should fix one of the unit test failures, looks like the `404` from the previous call to `h.ServeHTTP()` is being carried forward  in the recorder.

File: central/scannerdefinitions/handler/handler_test.go
Comment (dcaravel) [inline central/scannerdefinitions/handler/handler_test.go:431]:
```suggestion
	w = httptest.NewRecorder()
	req = s.getRequestVersion("dev")
```

File: central/scannerdefinitions/handler/handler_test.go
Comment (dcaravel) [inline central/scannerdefinitions/handler/handler_test.go:470]:
```suggestion
	w = httptest.NewRecorder()
	req = s.getRequestFile("name2repos")
```

**Review by dcaravel — APPROVED** [review]
LGTM after fixing the failing unit tests (added a few suggestions that I believe should fix the tests)

File: central/scannerdefinitions/handler/handler_test.go
Comment (RTann) [reply central/scannerdefinitions/handler/handler_test.go:403]: I did this to myself: https://github.com/stackrox/stackrox/pull/12210 :(

File: central/scannerdefinitions/handler/handler.go
Comment (RTann) [reply central/scannerdefinitions/handler/handler.go]: done here plus two other similar places

**Review by RTann — APPROVED** [review]

**Pull request merged**
